### DISCUSSION TITLE
Exporting color palette and lush spec in a single lua table

### DISCUSSION THREAD

**Discussion by nxmxsxs** [rktjmp/lush.nvim#86] [Q&A] [answered]
Hi, I was wondering if there's a metamethod magic I could apply to export a color palette and a lush spec in a single table. Thanks.

**Comment by rktjmp** [comment, marked as answer]
The lush spec itself is a table, so you should be able to do something like

```lua
local colorscheme = require("my-lush-theme")
-- then either access the color, convert it to a string or get the hex value directly
local bg = colorscheme.Normal.bg
bg.darken(10) -- can still apply operators
bg.h -- or access h,s,l, anything values
local bg_hex = colorscheme.Normal.bg.hex -- get as hex string
local str = tostring(colorscheme.Normal.bg) -- anything tostring'ing will convert to hex automatically
local rgb = colorscheme.Normal.bg.rgb -- {r, g, b} table (just added, so you may need to git pull)
```

If you want to store a normal table (or any other data), you can use the `lush` key which is retained in the spec:

https://github.com/rktjmp/lush.nvim/blob/87e9039138051ae75f76235924a273ac1a78cc3e/doc/lush.txt#L304-L306

Melange does this by storing the palette on an arbitrary group:

https://github.com/savq/melange/blob/ede563c0ed19c821d42006bdacdbb3515620865e/lua/melange/init.lua#L403

Which is then accessible via `colorscheme.Melange.lush`. 

You can also look at https://github.com/rktjmp/lush.nvim/blob/main/BUILD.md which lets you export your spec in any way you want.

**Reply by nxmxsxs** [reply]
I may have figured out a solution that works for me.
I export the lush spec as the first item in a tuple and color palette as the second item.
lush(lua) inherently ignores the second item so this wouldn't break `Lushify` for instance.
![lush_1](https://user-images.githubusercontent.com/95099980/162903311-c059ef72-9a45-497f-9fdc-17b40cf44ed0.png)
![lush_2](https://user-images.githubusercontent.com/95099980/162903393-7edbadb2-7e41-4e4c-9308-e297f1f91811.png)

**Reply by nxmxsxs** [reply]
nvm, I lied. The second item in the tuple would be nil when requiring.
I guess this makes sense because how would `package.loaded` cache "tuples"

**Comment by nxmxsxs** [comment]
I'll take a look at shipwright as I would like to export a palette for awesomewm

**Reply by rktjmp** [reply]
For what it's worth, I just add my theme to awesomewm's lua package.path and require the lush theme directly. Then I can just set my colors directly from highlight groups. Easy if you don't plan to share the theme.

**Reply by nxmxsxs** [reply]
I wouldn't want to awesomewm config to depend entirely on my neovim config and package manager so I think it might be better to generate a lua file containing the palette and place it directly in awesomewm config folder. Anyways, thanks for your help.

**Comment by nxmxsxs** [comment]
Here's what I was able to come up with

https://user-images.githubusercontent.com/95099980/163183058-3ef57b66-05a4-442e-b6fb-bcdac1f2b2c6.mov

**Reply by nxmxsxs** [reply]
I just noticed that there might be some slight errors with the conversion from hsl to hex.
For instance: `hsl('#F28FAD')` gets converted to `#F28DAB`, but some hex colors are converted correctly, example:
`hsl('#F5C2E7')` gets converted correctly to `#F5C2E7`. Is this expected behaviour or am I missing something?

**Reply by rktjmp** [reply]
It's an effect of moving between colorspaces, not really avoidable from what I have looked at. See #74 #68 #59. You can "avoid" it by defining your colours via HSL/HSLuv but since you seem to be patching an existing theme that might not be so easy for you.

As I describe in https://github.com/rktjmp/lush.nvim/issues/74, you can manually tweak the values to get a better match if they are too far out. You can apply the HSL operators inside the spec (before converting to hex) to modify the hex values from catppuchino if they need a little tweaking.

**Reply by nxmxsxs** [reply]
Would using floats as parameters in the hsl function increase the accuracy of the conversion or would there be little to no effect?

**Reply by rktjmp** [reply]
No effect, the input values are rounded to integers.

**Reply by nxmxsxs** [reply]
I meant if the parameters weren't rounded to integers, do you think it would have an effect?

**Reply by rktjmp** [reply]
The values will be different but probably only because they will be rounded somewhere else unintentionally as colorspaces are converted and they roll over integer values. If you're already working with HSL though you can just bump them by whole integers until you get a visual match as described in #74.

There aren't many round calls in Lush, you could experiment with disabling them.
